### ISSUE TITLE
Investigate Vivaldi-GN or similar approach to minimize GN patching

### ISSUE THREAD

**Issue by iefremov** [brave/brave-browser#15628]
Vivaldi crafted their own version of GN that simplifies adding sources/deps/etc to existing chromium GN targets. We could investigate similar approach to minimize patching.

see https://github.com/vivaldi/Vivaldi-GN/blob/master/docs/Vivaldi-GN.md

also see examples in Vivaldi source code (https://vivaldi.com/source/). Some of them are not quite clean because of circular dependencies, but we can see the idea.

```
# Copyright (c) 2015-2017 Vivaldi Technologies AS. All rights reserved
update_target("//chrome/browser:browser") {
  public_deps += [
    "//vivaldi/calendar",
  ]
  allow_circular_includes_from += [
    "//vivaldi/calendar",
  ]
}
update_target("//chrome:browser_dependencies") {
  public_deps += [ "//vivaldi/calendar" ]
}
```

**Comment by iefremov** [comment]
cc @bridiver @mariospr @mkarolin

**Comment by iefremov** [comment]
First steps would be:
1) check how the opensourced version corresponds to today's GN, whether it is easy to rebase
2) try to run it on a simple example
3) try to convert our existing `source.gni` to vivaldi-gn format

**Comment by mariospr** [comment]
I've been taking a look at this today and I have some comments:

  * Vivaldi's version of GN is now 3 years old: it's version `1458` from Dec 6th 2018 (vs version `1938` from Sep 23 2021, used in Chromium 96.0.4664.45).
  * Upstream GN already includes a bunch of patches by Vivaldi's developers yet, **doesn't include 4 patches** which are the most relevant ones for the overrides :
    - bebf13b5 [GN: Add support of alias paths in labels](https://github.com/vivaldi/Vivaldi-GN/commit/bebf13b5)
    - c8aede71 [GN: Add declare_overrides function.](https://github.com/vivaldi/Vivaldi-GN/commit/c8aede71)
    - 956b9bb0 [GN: Add updating of targets and template instances](https://github.com/vivaldi/Vivaldi-GN/commit/956b9bb0)
    - 7bf2c8ca [GN: Fix use of import in BUILDCONFIG.gn](https://github.com/vivaldi/Vivaldi-GN/commit/7bf2c8ca)
  * Worth noting that Vivaldi already trying upstreaming these 4 patches, but were not accepted:
    - For bebf13b5: https://chromium-review.googlesource.com/c/chromium/src/+/979441 and https://gn-review.googlesource.com/c/gn/+/2600
    - For c8aede71: https://chromium-review.googlesource.com/c/chromium/src/+/979440
    - For 956b9bb0: https://chromium-review.googlesource.com/c/chromium/src/+/979446
    - For 7bf2c8ca: https://chromium-review.googlesource.com/c/chromium/src/+/998892
  * I started applying such patches on top of GN version 1938 for now, but it's been far from trivial for now, in part probably because of me not being familiar neither with GN's codebase nor with Vivaldi's modifications.
    - Also worth noting that the source tree structure changed a bit and that also makes it difficult to apply the patches.
    - In any case, most bits seem to still apply, so I uploaded a WIP branch here: https://github.com/mariospr/gn/commits/Vivaldi-96.0.4664.45
      - It's not compiling properly for now due to some wrongly resolved conflicts

I'll try to keep looking into this on spare cycles if possible, first by trying to make the rebase of Vivaldi patches on top of GN 1938 compile and work (i.e. run tests), and then we'll be able to try to experiment with Brave's GN files.

**Comment by bridiver** [comment]
@mariospr it looks like at least some of those weren't accepted because gn was moved to a separate repo, did you check there?

**Comment by mariospr** [comment]
@bridiver Yes, I've actually based the work I did so far on that repo, and double-checked that those patches were not there.

It's true that most of the CLs I linked there are still on the old repo, though, but if you read the comments you'll see the patches have indeed not been accepted. Actually, there's one comment on the GErrit for the new repo stating that too: https://gn-review.googlesource.com/c/gn/+/2600/2#message-55eef6c7c8a1de1c2dfa11a84a97dda69949349f

**Comment by goodov** [comment]
I was experimenting with a patchless solution for gn changes and [created one](https://github.com/brave/brave-core/commit/3c3dcc36ea5af98f12505c551496fa4754e13e9b) that works with the original upstream `gn` binary and uses `chromium_src`-like approach. It works on `gn gen` phase and for each target looks for `chromium_src` counterpart in the format of `//brave/chromium_src/component/path/BUILD.<target_name>.gni` file which can contain changes for `sources, deps, public_deps, etc.`. The `gn gen` step on my machine went from 8-10 secs to around 20 secs, but that's basically nothing. The main issue with it is that it requires some validation on top, because any change can be easily lost without `gn` noticing it, but it may be solved in `check_chromium_src.py`-like manner.

@mariospr I also decided to dig a little into Vivaldi's approach and looked into their sources. To my surprise, Vivaldi releases everything, including `gn` sources with their changes. `vivaldi-source_5.0.2497.tar/vivaldi-source/thirdparty/gn` contains all things they describe on GitHub page, `gn` version they have is around Aug-Sept 2021.
